construction_impl.go

```diff
@@ -15,70 +15,318 @@
 
+func signSecp256k1(msg []byte, pk []byte) ([]byte, error) {
+	b2sum := blake2b.Sum256(msg)
+	sig, err := c.Sign(pk, b2sum[:])
+	if err != nil {
+		return nil, err
+	}
+
+	return sig, nil
+}
+
+// REVIEW: Doesn't actually verify the signature...
+// https://github.com/filecoin-project/lotus/blob/master/lib/sigs/secp/init.go#L38-L55
+func verifySecp256k1(sig []byte, a address.Address, msg []byte) error {
+	b2sum := blake2b.Sum256(msg)
+	pubk, err := c.EcRecover(b2sum[:], sig)
+	if err != nil {
+		return err
+	}
+
+	maybeaddr, err := address.NewSecp256k1Address(pubk)
+	if err != nil {
+		return err
+	}
+
+	if a != maybeaddr {
+		return fmt.Errorf("signature did not match")
+	}
+
+	if c.Verify(pubk, b2sum[:], sig) {
+		return nil
+	}
+
+	return fmt.Errorf("Invalid signature")
+}
+
 func (r RosettaConstructionFilecoin) DeriveFromPublicKey(publicKey []byte) (string, error) {
 	addr, err := address.NewSecp256k1Address(publicKey)
 	if err != nil {
 		return "", err
 	}
 	// FIXME: go=address does not allow setting a network
+	// https://github.com/filecoin-project/go-address/issues/6
 
 	return addr.String(), nil
 }
 
-func (r RosettaConstructionFilecoin) Sign(message []byte, privateKey []byte) ([]byte, error) {
-	signature, err := sigs.Sign(crypto.SigTypeSecp256k1, privateKey, message)
+func (r RosettaConstructionFilecoin) Sign(message []byte, sk []byte) ([]byte, error) {
+	return signSecp256k1(message, sk)
+}
+
+func (r RosettaConstructionFilecoin) Verify(message []byte, publicKey []byte, signature []byte) error {
+	addr, err := address.NewSecp256k1Address(publicKey)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return signature.MarshalBinary()
+	return verifySecp256k1(signature, addr, message)
 }
 
-func (r RosettaConstructionFilecoin) Verify(message []byte, publicKey []byte, signature []byte) error {
-	sig := crypto.Signature{}
-	err := sig.UnmarshalBinary(signature)
+func (r RosettaConstructionFilecoin) ConstructPayment(request *PaymentRequest) (string, error) {
+	to, err := address.NewFromString(request.To)
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	addr, err := address.NewSecp256k1Address(publicKey)
+	from, err := address.NewFromString(request.From)
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	return sigs.Verify(&sig, addr, message)
-}
+	value := types.NewInt(request.Quantity)
 
-func (r RosettaConstructionFilecoin) ConstructPayment(request *PaymentRequest) (string, error) {
-	panic("implement me")
+	gasprice, err := types.BigFromString(request.Metadata.GasPrice)
+	if err != nil {
+		return "", err
+	}
+
+	msg := &types.Message{types.MessageVersion,
+		to,
+		from,
+		request.Metadata.Nonce,
+		value,
+		gasprice,
+		request.Metadata.GasLimit,
+		builtin.MethodSend,
+		make([]byte,0),
+	}
+
+	tx, err := json.Marshal(msg)
+	if err != nil {
+		return "", err
+	}
+
+	return string(tx), nil
 }
 
 func (r RosettaConstructionFilecoin) ConstructMultisigPayment(request *MultisigPaymentRequest) (string, error) {
```

Comment:
`ConstructMultisigPayment` only create a proposition it need after to be approved. This should be a function that Coinbase will need. It is also a more complex process that might not fit in their workflow from what I have seen.

Reply:
I have opened an issue to track this:
https://github.com/Zondax/rosetta-filecoin-lib/issues/7